src/Resources/contao/dca/tl_user.php

```diff
@@ -985,4 +1019,155 @@ public function toggleVisibility($intId, $blnVisible, DataContainer $dc=null)
+
+		if ($activeRecord === null)
+		{
+			return;
+		}
+
+		foreach ($GLOBALS['TL_DCA']['tl_user']['palettes'] as $palette => $v)
+		{
+			if ($palette === '__selector__')
+			{
+				continue;
+			}
+
+			// Don't show 2FA options, if it is not the logged in user itself and 2FA is disabled.
+			if ($dc->id !== $this->User->id && (int) $activeRecord->use2fa !== '1')
+			{
+				continue;
+			}
+
+			$GLOBALS['TL_DCA']['tl_user']['palettes'][$palette] = str_replace
+			(
+				',password',
+				',password;{2fa_legend},use2fa',
+				$GLOBALS['TL_DCA']['tl_user']['palettes'][$palette]
+			);
+		}
+
+		// Only add subpalette if the logged in user is editing his own record
+		if ((int) $dc->id === (int) $this->User->id)
+		{
+			// extend selector
+			$GLOBALS['TL_DCA']['tl_user']['palettes']['__selector__'][] = 'use2fa';
+			$GLOBALS['TL_DCA']['tl_user']['subpalettes']['use2fa'] = '2faQrCode';
+
+			if ($this->User->confirmed2fa !== '1')
+			{
+				$GLOBALS['TL_DCA']['tl_user']['subpalettes']['use2fa'] .= ',confirmed2fa';
+			}
+		}
+	}
+
+	/**
+	 * Generates a 2FA secret if not present.
+	 *
+	 * @param DataContainer $dc
+	 *
+	 * @return string
+	 */
+	protected function generate2faSecret(DataContainer $dc)
+	{
+		$secret = $dc->activeRecord->secret;
+
+		if ($secret === null)
+		{
+			// Generate 1024 bit secret
+			$secret = random_bytes(128);
+
+			$this->Database->prepare("UPDATE tl_user SET secret=? WHERE id=?")
+				->execute($secret, $dc->id);
+
+			// Reload user with new secret
+			$this->User = BackendUser::getInstance();
+		}
+
+		return $secret;
+	}
+
+	/**
+	 * Save callback for 2FA.
+	 *
+	 * @param $varValue
+	 * @param DataContainer $dc
+	 *
+	 * @return mixed
+	 */
+	public function save2faSecret($varValue, DataContainer $dc)
+	{
+		$this->generate2faSecret($dc);
+
+		return $varValue;
+	}
+
+	/**
+	 * input_field_callback to display the QR code in the subpalette.
+	 *
+	 * @param DataContainer $dc
+	 *
+	 * @return string
+	 */
+	public function get2faQrCode(DataContainer $dc)
+	{
+		$secret = $this->generate2faSecret($dc);
+
+		if ($secret === '' || $secret === null)
+		{
+			return '';
+		}
+
+		/** @var \Contao\CoreBundle\Security\TwoFactor\ContaoTwoFactorAuthenticatorInterface $twoFactorAuthenticator */
+		$twoFactorAuthenticator = System::getContainer()->get('contao.security.two_factor.authenticator');
+
+		/** @var \Symfony\Component\HttpFoundation\Request $request */
+		$request = System::getContainer()->get('request_stack')->getCurrentRequest();
+
+		$class = $GLOBALS['TL_DCA'][$dc->table]['fields'][$dc->field]['eval']['tl_class'] . ' 2fa-qr-code';
+
+		return '
+<div class="' . $class . ' widget">
+  <div id="ctrl_' . $dc->field . '" class="">
+    <h3><label for="ctrl_' . $dc->field . '">' . $GLOBALS['TL_LANG']['tl_user']['2faQrCode'][0] . '</label></h3>
+    <img src="data:image/svg+xml;base64,' . base64_encode($twoFactorAuthenticator->getQrCode($this->User, $request)) . '" />
```

Comment:
Can we use `BackendUser::getInstance()` everywhere instead of `$this->User` that's stored somewhere in nowhere?